VSR: get rid of awaiting_checkpoint sync state

sentientwaffle opened the pull request:
Building on https://github.com/tigerbeetle/tigerbeetle/pull/2457.

(The first commit in this PR is #2457's squashed and rebased commits.)

---

Previously, state sync required receiving _two_ SV messages:

- the first SV message triggers grid cancelation, which is asynchronous, and
- the second SV message, received after cancelation completes, does the
  actual sync work.

This adds considerable delay, especially since the replica doesn't sync tables while in `awaiting_checkpoint`.

To fix this, make `on_start_view` asynchronous: if a replica receives SV that requires state sync, it _stores_ this particular SV message, and resumes its processing after cancelation is done.

<details>
<summary>Trace</summary>

In this trace, the blue bars each ~5s of `awaiting_checkpoint`, on a replica that is falling behind a (rapidly committing/checkpointing) cluster:

![chrome___tracing_](https://github.com/user-attachments/assets/6e2da1e3-5d33-413c-8f58-973220b6337a)

</details>

matklad committed:
vsr: make state-sync uninterruptible

This tweaks state sync to improve availability.

Previously, state sync required receiving _two_ SV messages:

- the first SV message triggers cancelation, which asynchronous
- the second SV message, received after cancelation completes, does the
  actual sync work

Notably, if a view change happens after the first message is received,
the state sync could be delayed. And, as the seed demonstrates, it could
even lead to the cluster getting stuck:

- The new primary doesn't necessary know the old primary's commit_max
- So it's SV might not be enough to complete state sync triggered by an
  SV from the previous primary.

To fix this, make `on_start_view` asynchronous: if a replica receives SV
that requires state sync, it _stores_ this particular SV message, and
resumes its processing after cancelation is done.

Seed: ./zig/zig build -Drelease vopr -- 13226853613775535404

sentientwaffle committed:
VSR: Make state-sync uninterruptible (part 2)

The `nonce mismatch` check is moved up out of `ignore_view_change_message` because otherwise:

1. While in status=recovering_head, receive a SV message from a newer view (with nonce=0).
2. Since `message.header.view > self.view` we accept it and set `self.view = message.header.view`.
3. Then we call `on_start_view_set_journal(message)` (still with that same SV).
4. Only now the views match, so we require that the nonces match as well -- we panic on `assert(!self.ignore_view_change_message(message.base_const()))`.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -2211,13 +2228,34 @@ pub fn ReplicaType(
             assert(message.header.command == .start_view);
             if (self.ignore_view_change_message(message.base_const())) return;
 
+            if (self.status == .recovering_head) {
+                if (message.header.view > self.view or
+                    message.header.op >= self.op_prepare_max() or
+                    message.header.nonce == self.nonce)
+                {
+                    // This SV is guaranteed to have originated after the replica crash,
+                    // it is safe to use to determine the head op.
+                } else {
+                    log.mark.debug(
+                        "{}: on_start_view: ignoring (recovering_head, nonce mismatch)",
+                        .{self.replica},
+                    );
+                    return;
+                }
+            }
```
I pulled the `nonce mismatch` check up out of `ignore_view_change_message()` because otherwise:

1. While in `status=recovering_head`, receive a SV message from a newer view (with nonce=0).
2. Since `message.header.view > self.view` we accept it and set `self.view = message.header.view`.
3. Then we call `on_start_view_set_journal(message)` (still with that same SV).
4. Only now the views match, so we require that the nonces match as well -- we panic on `assert(!self.ignore_view_change_message(message.base_const()))`.

matklad commented on src/vsr/replica.zig:
```diff
@@ -2272,75 +2337,69 @@ pub fn ReplicaType(
+            // Cluster is on the next checkpoint, and that checkpoint is durable and is safe to
+            // sync to. Try to optimistically avoid state sync and prefer WAL repair, unless
+            // there's evidence that the repair can't be completed.
+            const likely_stuck = self.syncing == .idle and self.repair_stuck();
```
This I think would be the right place to implement that other idea where a replica which syncs tables triggers state sync more aggressively, to avoid the situation where it waits for an entire `repair_sync_timeout` before triggering the sync, all while the cluster is making progress in the new checkpoint already.

sentientwaffle replied:
It is the right place, but I think I will include the actual change in https://github.com/tigerbeetle/tigerbeetle/pull/2621 instead. The reason is that as of that PR, `sync_tables` will get set earlier (immediately after we sync the checkpoint, rather than after state-machine open), so checking `sync_tables != null` can be used to answer the "are we already busy syncing" question.

sentientwaffle committed:
Journal: Refactor write_prepare_header

sentientwaffle committed:
Journal: Fix write race

Scenario:

1. `Journal.set_header_as_dirty(op=114 view=1)`
2. `Journal.write_prepare(op=114 view=1)`
3. `Journal.set_header_as_dirty(op=114 view=2)`
4. `Journal.write_prepare(op=114 view=2)`
5. `Journal.set_header_as_dirty(op=114 view=1)`
6. Finish writing the prepare/header of `op=114 view=1`. Since this matches our `journal.headers[slot]`, we mark the slot as clean.
7. Finish writing the prepare of `op=114 view=2`. We set `prepare_checksums` (in `write_prepare_header()`), but don't write the header because `entry changed while writing headers`.

At this point, the header's entry is (incorrectly!) clean. This is soon detected by an assert in `Journal.has_prepare()`, since it is clean but `journal.prepare_checksums` doesn't match `journal.headers`.

Seed: ./zig/zig build -Drelease vopr -- 4147204242601358017

sentientwaffle commented:
https://github.com/tigerbeetle/tigerbeetle/pull/2629/commits/ac89339d5a0a104c15e938f8617bd6dcd2d26136 is (afaict) a latent journal bug. I'm not 100% clear why we didn't discover it before -- I guess this PR allows view changes to be faster in some circumstances.

Also interesting to note that vopr found it, but not vopr-lite, even after many seeds.

matklad commented on src/vsr/replica.zig:
```diff
@@ -1388,6 +1395,18 @@ pub fn ReplicaType(
+                    return;
+                },
+                .updating_checkpoint => {},
+            }
```
TBH, this feels like a cludge to me, as it violates "embrace concurrency" principle. But also, a replica in the middle of `on_start_view` is in a rather weird state, and carefully handling that was hard.

matklad reviewed (approved):
LGTM-ish!

I am still not happy with the overall flow here, it feels complicated, but it's not I was able to come up with anything better while I was sitting on this code!
